src/purple_vm/src/primitives/value.rs

```diff
@@ -82,6 +85,96 @@ impl VmValue {
             _ => panic!(),
         }
     }
+
+    pub fn is_positive(&self) -> bool {
+        match *self {
+            VmValue::I32(val) => val >= 0,
+            VmValue::I64(val) => val >= 0,
+            VmValue::F32(val) => val >= 0.0,
+            VmValue::F64(val) => val >= 0.0,
+            VmValue::i32Array2(val) => val.iter().all(|&v| v >= 0),
+            VmValue::i32Array4(val) => val.iter().all(|&v| v >= 0),
+            VmValue::i32Array8(val) => val.iter().all(|&v| v >= 0),
+            VmValue::i64Array2(val) => val.iter().all(|&v| v >= 0),
+            VmValue::i64Array4(val) => val.iter().all(|&v| v >= 0),
+            VmValue::i64Array8(val) => val.iter().all(|&v| v >= 0),
+            VmValue::f32Array2(val) => val.iter().all(|&v| v >= 0.0),
+            VmValue::f32Array4(val) => val.iter().all(|&v| v >= 0.0),
+            VmValue::f32Array8(val) => val.iter().all(|&v| v >= 0.0),
+            VmValue::f64Array2(val) => val.iter().all(|&v| v >= 0.0),
+            VmValue::f64Array4(val) => val.iter().all(|&v| v >= 0.0),
+            VmValue::f64Array8(val) => val.iter().all(|&v| v >= 0.0),
+            _ => panic!(),
+        }
+    }
+
+    fn check_f32_infinite(val: f32) -> Option<f32> {
+        if val.is_infinite() {
+            None
+        } else {
+            Some(val)
+        }
+    }
+
+    fn check_f64_infinite(val: f64) -> Option<f64> {
+        if val.is_infinite() {
+            None
+        } else {
+            Some(val)
+        }
+    }
+
+    fn sum_f32(val1: &f32, val2: &f32) -> Option<f32> {
+        VmValue::check_f32_infinite(val1 + val2)
+    }
+
+    fn sum_f64(val1: &f64, val2: &f64) -> Option<f64> {
+        VmValue::check_f64_infinite(val1 + val2)
+    }
+
+    fn sub_f32(val1: &f32, val2: &f32) -> Option<f32> {
+        VmValue::check_f32_infinite(val1 - val2)
+    }
+
+    fn sub_f64(val1: &f64, val2: &f64) -> Option<f64> {
+        VmValue::check_f64_infinite(val1 - val2)
+    }
+
+    fn mul_f32(val1: &f32, val2: &f32) -> Option<f32> {
+        VmValue::check_f32_infinite(val1 * val2)
+    }
+
+    fn mul_f64(val1: &f64, val2: &f64) -> Option<f64> {
+        VmValue::check_f64_infinite(val1 * val2)
+    }
+
+    fn div_f32(val1: &f32, val2: &f32) -> Option<f32> {
+        if *val2 == 0.0 {
+            panic!("Attempted to divide by zero!")
```

Comment:
This shouldn't panic here. Return `None` instead propagate errors upward. What if val2 is calculated to be 0 at runtime?